### PR TITLE
uncompress kernel

### PR DESCRIPTION
**Pull request by rawdaGastan**
### Related Issues

- https://github.com/threefoldtech/zos/issues/2236

### CODE REVIEW

File: pkg/primitives/vm/decompress_kernel.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	"os"
+	"os/exec"
+
+	"github.com/cyberdelia/lzo"
+	"github.com/klauspost/compress/zstd"
+	"github.com/pierrec/lz4/v4"
+	"github.com/ulikunitz/xz"
+)
+
+func isValidELFKernel(KernelImagePath string) error {
```
Comment (xmonader) [inline pkg/primitives/vm/decompress_kernel.go:22]: I think we can use debug/elf https://pkg.go.dev/debug/elf

File: pkg/primitives/vm/decompress_kernel.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+func unZstd(kernelStream io.Reader) (io.Reader, error) {
+	r, err := zstd.NewReader(kernelStream)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func tryDecompressKernel(KernelImagePath string) error {
```
Comment (muhamadazmy) [inline pkg/primitives/vm/decompress_kernel.go]: why argument name is starting with an UpperCase ?

File: pkg/primitives/vm/decompress_kernel.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+			name:           "zstd",
+		},
+	}
+
+	var errs error
+
+	for _, algo := range algos {
+		err = decompressData(kernelData, tmpFile, algo)
+		if err == nil {
+			f, err := os.Create(KernelImagePath)
```
Comment (muhamadazmy) [inline pkg/primitives/vm/decompress_kernel.go]: you need to close this file

File: pkg/primitives/vm/decompress_kernel.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+		return fmt.Errorf("kernel image is required")
+	}
+
+	// if kernel is already an elf (uncompressed)
+	if err := isValidELFKernel(KernelImagePath); err == nil {
+		return nil
+	}
+
+	// Prepare temp files:
+	tmpFile, err := os.CreateTemp("/tmp", "vmlinux-")
```
Comment (muhamadazmy) [inline pkg/primitives/vm/decompress_kernel.go]: according to the docs. you should set `dir` (first argument) to empty string `""` to use the default tmp directory. It's better than setting it up explicitly imho

File: pkg/primitives/vm/decompress_kernel.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+	if err != nil {
+		return err
+	}
+	defer tmpFile.Close()
+	defer os.Remove(tmpFile.Name())
```
Comment (muhamadazmy) [inline pkg/primitives/vm/decompress_kernel.go]: The close has to happen first then deleting right ?
so this has to be in the opposite order. 

Or better 

```suggestion
        defer func() {
            tmpFile.Close()
            or.Remove(tmpFile.Name())
        }
```

File: pkg/primitives/vm/utils.go [hunk 6]
```diff
@@ -148,6 +148,12 @@ func (p *Manager) prepContainer(
 
 	// can be overridden from the flist itself if exists
 	if len(imageInfo.KernelPath) != 0 {
+		// try to decompress kernel
+		err := tryDecompressKernel(imageInfo.KernelPath)
+		if err != nil {
```
Comment (muhamadazmy) [inline pkg/primitives/vm/utils.go]: Just some cosmetics but i prefer always to do:

```go 
if err := functionThatMightFail(); err != nil {
   // handle error
}
```
Go error handling is already annoying, so this makes it a little bit compact

File: pkg/primitives/vm/decompress_kernel.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+		{
+			decompressFunc: unlzma,
+			headerBytes:    []byte("\135\000\000\000"),
+			name:           "unlzma",
+		},
+		// {
+		// 	decompressFunc: lZop,
+		// 	headerBytes:    []byte("\211\114\132"),
+		// 	name:           "lzop",
+		// },
```
Comment (muhamadazmy) [inline pkg/primitives/vm/decompress_kernel.go]: This function should still be there, with an implementation that only return error. But we can then return a clear error that we don't support lzop

**Review by muhamadazmy — CHANGES_REQUESTED** [review]

**Review by muhamadazmy — APPROVED** [review]

**Pull request merged**